pkg/network/ebpf/c/prebuilt/tracer.c

```diff
@@ -834,6 +843,100 @@ int kretprobe__inet6_bind(struct pt_regs* ctx) {
     return sys_exit_bind(ret);
 }
 
+SEC("kprobe/sockfd_lookup_light")
+int kprobe__sockfd_lookup_light(struct pt_regs* ctx) {
+    int sockfd = (int)PT_REGS_PARM1(ctx);
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+
+    // Check if have already a map entry for this pid_fd_t
+    pid_fd_t key = {
+        .pid = pid_tgid >> 32,
+        .fd = sockfd,
+    };
+    struct sock** sock = bpf_map_lookup_elem(&sock_by_pid_fd, &key);
+    if (sock != NULL) {
+        return 0;
+    }
+
+    bpf_map_update_elem(&sockfd_lookup_args, &pid_tgid, &sockfd, BPF_ANY);
+    return 0;
+}
+
+// this kretprobe is essentially creating:
+// * an index of pid_fd_t to a struct sock*;
+// * an index of struct sock* to pid_fd_t;
+SEC("kretprobe/sockfd_lookup_light")
+int kretprobe__sockfd_lookup_light(struct pt_regs* ctx) {
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    int *sockfd = bpf_map_lookup_elem(&sockfd_lookup_args, &pid_tgid);
+    if (sockfd == NULL) {
+        return 0;
+    }
+
+    // For now let's only store information for TCP sockets
+    struct socket* socket = (struct socket*)PT_REGS_RC(ctx);
+    enum sock_type sock_type = 0;
+    bpf_probe_read(&sock_type, sizeof(short), &socket->type);
+    if (sock_type != SOCK_STREAM) {
+        goto cleanup;
+    }
+
+    // Retrieve struct sock* pointer from struct socket*
+    struct sock *sock = NULL;
+    bpf_probe_read(&sock, sizeof(sock), (char*)socket + offset_socket_sk());
+
+    pid_fd_t pid_fd = {
+        .pid = pid_tgid >> 32,
+        .fd = (*sockfd),
+    };
+
+    // These entries are cleaned up by tcp_close
+    bpf_map_update_elem(&pid_fd_by_sock, &sock, &pid_fd, BPF_ANY);
+    bpf_map_update_elem(&sock_by_pid_fd, &pid_fd, &sock, BPF_ANY);
+cleanup:
+    bpf_map_delete_elem(&sockfd_lookup_args, &pid_tgid);
+    return 0;
+}
+
+SEC("kprobe/do_sendfile")
+int kprobe__do_sendfile(struct pt_regs* ctx) {
+    u64 fd_out = (int)PT_REGS_PARM1(ctx);
+    u64 fd_in = (int)PT_REGS_PARM2(ctx);
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    u64 val = (fd_in << 32) | fd_out;
```

Comment:
Do we still need both `fd`s, since we are only using `fd_out`?

Reply:
Only `fd_out`. I'll update it